### ISSUE TITLE
Benchmark with tensor network contraction orders

### ISSUE THREAD

**Issue by GiggleLiu** [JuliaQX/QXTools.jl#32]
Nice package! I saw you implemented FlowCutter in this repo, but I am not quite familiar with it. Maybe we can benchmark it with existing contraction order finders.

https://github.com/TensorBFS/OMEinsumContractionOrders.jl

I feel the `TreeSA` method in it is one of the state of the art.

**Comment by brenjohn** [comment]
Thanks @GiggleLiu :)

Yeah, I agree. It would be interesting to benchmark these methods. One thing I'm wondering about is which test cases to use to get a meaningful benchmark. We could compare them on tensor networks representing the circuits found in the qflex repo for instance:

https://github.com/ngnrsaa/qflex/tree/master/config/circuits

Although my concern is these may not be very representative of circuits designed for use cases other than benchmarking hardware. Do you have any thoughts on this or suggestions? I can compare them on this test set otherwise :)

**Comment by GiggleLiu** [comment]
Thanks for the info. I did not mean we should spend a lot of time on it to do a serious benchmark, maybe in the future we might want to do so. But Now, I think it is good to start from some toy circuits, like any of the big enough circuits in the qflex repo. All should give fair performance measure of contraction order finder when they are large enough. So what about this one, it is quite popular https://github.com/ngnrsaa/qflex/blob/master/config/circuits/sycamore_53_20_0.txt 

If you can convert circuits from Yao, here we also have a google 53 qubit circuit: https://github.com/QuantumBFS/YaoExtensions.jl/blob/master/src/easybuild/google53.jl 

I can write a script for optimizing contraction orders using TreeSA and KaHyPar algorithms. Let me submit a PR later, or do you want to set up a new repo?

**Comment by brenjohn** [comment]
Yeah that sounds good. I think we do have some code to convert circuits from Yao so we can try both.

I should have some time tomorrow which I can use to find some orders for these with FlowCutter.

I think it'd be a bit cleaner to use a new repo so I set this one up for us:
https://github.com/brenjohn/Contraction-Order-Bench

**Comment by brenjohn** [comment]
I didn't get around to using the google circuit with Yao yet but I had some old code I could use earlier to convert the qflex qasm files. I ran flowcutter a couple of times on the line graph for the sycamore circuit with different parameters. 

The decompose variable indicates if two qubit gates are decomposed into rank 3 tensors before creating the line graph and the hypergraph variable indicates if the line graph was constructed from the circuits hypergraph or regular graph.

I'm not sure what you think the best way to compare orders would be but I looked at the treewidth of the results and it looks like these variables don't have much effect on the result in this case. FlowCutter seems to converge pretty quickly to a treewidth of 57 and then gets stuck there :D (I ran it once in the past for 30 minutes on this circuit and it didn't improve on 57). I think the best treewidth I've seen in the literature for this circuit is 53 which used a method similar to KaHyPar I think.  I'd be curious to see what `TreeSA` gives.

![flowcutter_sycamore_53_20_0](https://user-images.githubusercontent.com/27730596/140587009-f7bf649e-1ef7-43ec-8c28-5b9691a7846b.png)
